### PR TITLE
Transformations: Add substring matcher to the 'Filter by Value' transformation

### PR DESCRIPTION
**Pull request by timlevett**
**What is this feature?**

This feature adds in a new matcher type called "substring" and "not substring" that does substring matching on field values to filter results. See demo below:

https://github.com/grafana/grafana/assets/3534544/ccfc38f2-abac-44fb-8d13-74743e77547c

**Why do we need this feature?**

Being able to do partial string matching is a huge value for table data manipulation and combined with template variables being used in transformations, this makes sharable filtering a reality.

**Who is this feature for?**

Creators of dashboards that want to transform data

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.test.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
+import { toDataFrame } from '../../../dataframe';
+import { DataFrame } from '../../../types/dataFrame';
+import { getValueMatcher } from '../../matchers';
+import { ValueMatcherID } from '../ids';
+
+describe('value substring to matcher', () => {
```
Comment (nmarrs) [inline packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.test.ts:6]: ❤️

File: packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  id: ValueMatcherID.notSubstring,
+  name: 'Is not substring',
+  description: 'Match where value for given field is not a substring to options value.',
+  get: (options) => {
+    return (valueIndex: number, field: Field) => {
+      const value = field.values[valueIndex];
+      // eslint-disable-next-line eqeqeq
+
+      // compare value to options.value, if
+      return !value.includes(options.value);
```
Comment (nmarrs) [inline packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.ts:35]:
```suggestion
      return !value.includes(options.value);
```

File: packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+const isLikeValueMatcher: ValueMatcherInfo<BasicValueMatcherOptions> = {
+  id: ValueMatcherID.substring,
+  name: 'Is Substring',
+  description: 'Match where value for given field is a substring to options value.',
+  get: (options) => {
+    return (valueIndex: number, field: Field) => {
+      const value = field.values[valueIndex];
+      // eslint-disable-next-line eqeqeq
+      return value.includes(options.value);
```
Comment (nmarrs) [inline packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.ts:15]:
```suggestion
      return value.includes(options.value);
```

Comment (nmarrs) [reply packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.ts:15]: ^ it also seems like something is going wrong with this logic (value is null and is causing a runtime error in the transformation)

**Review by nmarrs — COMMENTED** [review]
Hey @timlevett! This is an awesome contribution! Code wise I think this is looking pretty good, left a few nit suggestions.

Functionality wise I am running into some issues getting it to work as expected locally

https://github.com/grafana/grafana/assets/22381771/6ec534cf-9014-425e-b783-31a0a83fc436


For some reason the field value is null and thus causing a runtime error in the transformation logic. Here is the test dashboard that showcases this issue

[Sub string filter test dashboard-1709074359514.json](https://github.com/grafana/grafana/files/14426516/Sub.string.filter.test.dashboard-1709074359514.json)

**Review by codeincarnate — APPROVED** [review]
Looks good and was working for me. Please see my comments 😄

**Comment by timlevett** [comment]
> For some reason the field value is null and thus causing a runtime error in the transformation logic. Here is the test dashboard that showcases this issue

Ah, i bet i need to check for not null, let me add that.

File: packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
+import { Field, FieldType } from '../../../types/dataFrame';
+import { ValueMatcherInfo } from '../../../types/transformations';
+import { ValueMatcherID } from '../ids';
+
+import { BasicValueMatcherOptions } from './types';
+
+const isLikeValueMatcher: ValueMatcherInfo<BasicValueMatcherOptions> = {
```
[the remark this answers is not in the document]
Comment (timlevett) [reply packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.ts:7]: ah, this was carryover from the initial creation where i named it "like", will fix

File: packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  id: ValueMatcherID.substring,
+  name: 'Is Substring',
```
[the remark this answers is not in the document]
Comment (timlevett) [reply packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.ts:9]: Can do, I used "is" as that is what we do for all the other matchers. "is equal", "is null", etc...

File: packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (timlevett) [reply packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.ts]: Fixed in 572aec8ebb7

File: packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    };
+  },
+  getOptionsDisplayText: () => {
+    return `Matches all rows where field is similar to the value.`;
+  },
+  isApplicable: (field) => field.type === FieldType.string,
+  getDefaultOptions: () => ({ value: '' }),
+};
+
+const isNotLikeValueMatcher: ValueMatcherInfo<BasicValueMatcherOptions> = {
```
[the remark this answers is not in the document]
Comment (timlevett) [reply packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.ts]: Fixed in 572aec8ebb7

File: packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (timlevett) [reply packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.ts]: Updated in acf078d6f38

**Comment by codeincarnate** [comment]
Thanks for the changes @timlevett 🙏, reviewing this PR made me surprised we didn't already have this. Awesome stuff! 😄

File: packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.test.ts (hunk 1 above)
Comment (baldm0mma) [reply packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.test.ts:6]: Teeeeeeeestssssss!!!

**Review by baldm0mma — APPROVED** [review]
Lovely!

**Comment by nmarrs** [comment]
Testing this locally in current state I'm still getting errors 😬 https://github.com/grafana/grafana/pull/83548#pullrequestreview-1904931540

but maybe its due to using testdata or something weird

**Comment by baldm0mma** [comment]
@timlevett after seeing @nmarrs issue, I retested with `null` values, and I am also hitting the same issue: `cannot read properties of null ('reading includes')`. I sounds like what you mentioned above: there needs to be some null check. 👍

**Comment by timlevett** [comment]
Fixed NPE issue in 67d58f93899

File: packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+};
+
+const isNotSubstringValueMatcher: ValueMatcherInfo<BasicValueMatcherOptions> = {
+  id: ValueMatcherID.notSubstring,
+  name: 'Does not contain substring',
+  description: 'Match where value for given field is not a substring to options value.',
+  get: (options) => {
+    return (valueIndex: number, field: Field) => {
+      const value = field.values[valueIndex];
+      return options.value !== '' && (!value || !value.includes(options.value));
```
Comment (leeoniya) [inline packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.ts:31]: `!value` wouldn't work for `'0'`, right?

**Comment by baldm0mma** [comment]
Null check doesn't seem to be an issue anymore, but using @nmarrs 's data above, I'm not seeming to be able to filter as expected. Trying to filter out "ford" but not filtering... 🤔

[Sub string filter test dashboard-1709225190437.json](https://github.com/grafana/grafana/files/14451044/Sub.string.filter.test.dashboard-1709225190437.json)

File: packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.ts (hunk 7 above)
Comment (timlevett) [reply packages/grafana-data/src/transformations/matchers/valueMatchers/substringMatchers.ts]: Fixed in https://github.com/grafana/grafana/pull/83548/commits/2f93ba25f71589e08539d4d1f3b54ef7fae684dc

**Comment by timlevett** [comment]
> Null check doesn't seem to be an issue anymore, but using @nmarrs 's data above, I'm not seeming to be able to filter as expected. Trying to filter out "ford" but not filtering... 🤔
> 
> [Sub string filter test dashboard-1709225190437.json](https://github.com/grafana/grafana/files/14451044/Sub.string.filter.test.dashboard-1709225190437.json)

I just saw this comment after i clicked merge,let me take a look and i'll follow up with a PR if needed.

**Comment by timlevett** [comment]
@baldm0mma I tried to use your dashboard-json but it was erring out on your datasource, saying it didn't exist, looking at the json its a csv? anyway, I setup another simple test and i can't replicate what you are experiencing on the latest commit. If you can reproduce on main let me know 🙏🏻 .

**Comment by baldm0mma** [comment]
@timlevett no worries! Tried again on `main`, and it's great! Thanks!!!

**Pull request merged**